src/Agent.Listener/Configuration/WindowsSecurityManager.cs

```diff
@@ -0,0 +1,477 @@
+                }
+            }
+
+            Trace.Verbose(StringUtil.Format("LogonUser returned with result {0}", result));
+            return result != 0;
+        }
+
+        public NTAccount GetDefaultServiceAccount()
+        {
+            SecurityIdentifier sid = new SecurityIdentifier(WellKnownSidType.LocalServiceSid, domainSid: null);
+            NTAccount account = sid.Translate(typeof(NTAccount)) as NTAccount;
+
+            if (account == null)
+            {
+                throw new InvalidOperationException(StringUtil.Loc("LocalServiceNotFound"));
+            }
+
+            // TODO: If its domain joined machine use WellKnownSidType.NetworkServiceSid?
+            return account;
+        }
+
+        public void SetPermissionForAccount(string path, string accountName)
+        {
+            Trace.Info(nameof(SetPermissionForAccount));
+
+            var windowsSecurityHandler = HostContext.GetService<IWindowsSecurityManager>();
+
+            try
+            {
+                string groupName = windowsSecurityHandler.GetUniqueBuildGroupName();
+
+                Trace.Info(StringUtil.Format("Calculated unique group name {0}", groupName));
+                if (!windowsSecurityHandler.LocalGroupExists(groupName))
+                {
+                    Trace.Info(StringUtil.Format("Trying to create group {0}", groupName));
+                    windowsSecurityHandler.CreateLocalGroup(groupName);
+                }
+
+                Trace.Info(StringUtil.Format("Trying to add user {0} to the group {0}", accountName, groupName));
+                windowsSecurityHandler.AddMemberToLocalGroup(accountName, groupName);
+
+                Trace.Info(StringUtil.Format("Set full access control to group for the folder {0}", path));
+                // TODO Check if permission exists, this could cause ACL not being canonical order when we set the permission for the second time
+                // windowsSecurityHandler.GrantFullControlToGroup(path, groupName);
+
+                Trace.Info("Verifying if the account has LogonAsService permission");
+
+                if (!windowsSecurityHandler.GrantUserLogonAsServicePrivilage(accountName))
+                {
+                    throw new InvalidOperationException(StringUtil.Format("Can not grant LogonAsService permission to the user {0}", accountName));
+                }
+            }
+            catch (Exception exception)
+            {
+                Trace.Error(exception);
+                throw new InvalidOperationException(StringUtil.Loc("CanNotSetPermissionForAccount"));
+            }
+        }
+
+        private static byte[] GetSidBinaryFromWindows(string domain, string user)
+        {
+            try
+            {
+                SecurityIdentifier sid = (SecurityIdentifier)new NTAccount(string.Format(domain + "\\" + user).TrimStart('\\')).Translate(typeof(SecurityIdentifier));
```

Comment:
same here, make sure you test all different domain/user combination, the current .net agent should have some bug around this area that i haven't fixed.